src/box/memtx_tree.cc

```diff
@@ -1156,9 +1165,104 @@ static ssize_t
 memtx_tree_index_count(struct index *base, enum iterator_type type,
 		       const char *key, uint32_t part_count)
 {
-	if (type == ITER_ALL)
-		return memtx_tree_index_size<USE_HINT>(base); /* optimization */
-	return generic_index_count(base, type, key, part_count);
+	assert((base->def->opts.hint == INDEX_HINT_ON) == USE_HINT);
+
+	struct region *region = &fiber()->gc;
+	RegionGuard region_guard(region);
+
+	struct memtx_tree_index<USE_HINT> *index =
+		(struct memtx_tree_index<USE_HINT> *)base;
+
+	if (canonicalize_lookup(base->def, &type, &key, part_count) == -1)
+		return -1;
+
+	memtx_tree_t<USE_HINT> *tree = &index->tree;
+	struct key_def *cmp_def = memtx_tree_cmp_def(&index->tree);
+	struct memtx_tree_key_data<USE_HINT> start_data;
+	start_data.key = key;
+	start_data.part_count = part_count;
+	if (USE_HINT)
+		start_data.set_hint(key_hint(key, part_count, cmp_def));
+	struct memtx_tree_key_data<USE_HINT> null_after_data = {};
+	memtx_tree_iterator_t<USE_HINT> unused1;
+	size_t begin_offset;
+	bool unused2;
+	struct memtx_tree_data<USE_HINT> *initial_elem;
+	if (!memtx_tree_lookup(tree, &start_data, null_after_data, &type,
+			       region, &unused1, &begin_offset, &unused2,
+			       &initial_elem))
+		return 0;
+
+	struct txn *txn = in_txn();
+	struct space *space = space_by_id(base->def->space_id);
+	size_t full_size = memtx_tree_size(tree);
+	size_t end_offset;
+
```

Comment:
For the case when index is unique, iterator is `EQ` or `REQ` and full key is specified, we could simply call `memtx_tree_index_get_internal` as an optimization and see at its returned value. We will conflict with updates then, but it seems to me that advantages are more significant:
1. We will use "local" trackers instead of "global" ones (`track_point`/`track_read`/`inplace_gap` instead of `nearby_gap`/`count_gap`).
2. We won't need to patch `memtx_tx_handle_gap_write` (that extra logic for equality iterators with full key specified).

Reply:
Made it use the `memtx_tx_track_point` function on missed full equal key lookup and removed the `memtx_tx_handle_gap_write` changes as not required anymore:

```diff
--- a/src/box/memtx_tree.cc
+++ b/src/box/memtx_tree.cc
@@ -1198,6 +1198,15 @@ memtx_tree_index_count(struct index *base, enum iterator_type type,
        size_t full_size = memtx_tree_size(tree);
        size_t end_offset;
 
+       /* Fast path: not found equal with full key. */
+       if (start_data.part_count == cmp_def->part_count &&
+           !equals && (type == ITER_EQ || type == ITER_REQ)) {
+/********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND START*********/
+               memtx_tx_track_point(txn, space, base, start_data.key);
+/*********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND END**********/
+               return 0; /* No tuple matching the full key. */
+       }
+
        /* Fast path: not found with reverse iterator. */
        if (begin_offset == size_t(-1)) {
                assert(iterator_type_is_reverse(type));
```

I don't like this invonsistency of conflicting cases for the `count` function, so the `GAP_COUNT` is still created, but if it will become an issue it can be fixed by simply moving the `COUNT` gap item into a specific element and introducing a little of additional logic based on it.

Please let me know if you're OK with it.

Reply:
It's created in the slow path (even for full key with EQ iterator).

Reply:
Thanks for optimization, looks nice!